lib/OpenQA/WebAPI/Controller/API/V1/Job.pm

```diff
@@ -738,5 +743,89 @@ sub whoami {
     $self->render(json => {id => $jobid});
 }
 
+=over 4
+
+=item _generate_job_setting()
+
+Create job for product matching the contents of the DISTRI, VERSION, FLAVOR and ARCH, MACHINE
+settings, and returns a job's settings. Internal method used in the B<create()> method.
+
+=back
+
+=cut
+
+sub _generate_job_setting {
+    my ($self, $args) = @_;
+    my $schema = $self->schema;
+
+    my %settings;    # Machines, product and test suite settings for the job
+    my @classes;     # Populated with WORKER_CLASS settings from machines and products
+
+    # Populated with Product settins if there are DISTRI, VERSION, FLAVOR, ARCH in arguments.
+    if (   defined $args->{DISTRI}
+        && defined $args->{VERSION}
+        && defined $args->{FLAVOR}
+        && defined $args->{ARCH})
+    {
+        my $products = $schema->resultset('Products')->search(
+            {
+                distri  => $args->{DISTRI},
+                version => $args->{VERSION},
+                arch    => $args->{ARCH},
+                flavor  => $args->{FLAVOR},
+            });
+
+        if (my $product = $products->next) {
+            my %tmp_setting = map { $_->key => $_->value } $product->settings;
+
+            if (my $class = delete $tmp_setting{WORKER_CLASS}) {
+                push @classes, $class;
+            }
+            @settings{keys %tmp_setting} = values %tmp_setting;
+        }
+    }
+
+    # Populated with machine settings if there is MACHINE in arguments.
+    if (defined $args->{MACHINE}) {
+        my $machines = $schema->resultset('Machines')->search(
+            {
+                name => $args->{MACHINE},
+            });
+
+        if (my $machine = $machines->next) {
+            my %tmp_setting = map { $_->key => $_->value } $machine->settings;
+
+            if (my $class = delete $tmp_setting{WORKER_CLASS}) {
+                push @classes, $class;
+            }
+            @settings{keys %tmp_setting} = values %tmp_setting;
+        }
+    }
+
+    # TEST is mandatory, so populate with TestSuite settings.
+    my $test_suites = $schema->resultset('TestSuites')->search(
+        {
+            name => $args->{TEST},
+        });
+
+    if (my $test_suite = $test_suites->next) {
+        my %test_suite_setting = map { $_->key => $_->value } $test_suite->settings;
+
+        if (my $test_suite_class = delete $test_suite_setting{WORKER_CLASS}) {
+            push @classes, $test_suite_class;
+        }
+        @settings{keys %test_suite_setting} = values %test_suite_setting;
+    }
+
+    $settings{WORKER_CLASS} = join ',', sort @classes if @classes > 0;
+
+    for (keys %$args) {
+        $settings{uc $_} = $args->{$_};
+    }
+
+    my $expand_result = OpenQA::ExpandPlaceholder::expand_placeholders(\%settings);
+    return {expand_result => $expand_result, settings_result => \%settings};
```

Comment:
Like in `_generate_jobs` the variable and hash key should be called `$error_message`. Otherwise the code on the call-site looks strange: `json => {error => $result->{expand_result}}` - why would one return the "result" as an error? This makes it also consistent with `_generate_jobs`.

Reply:
Thanks for reminding, I missed this file, I have modified it.